mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoImplementationGenerator.java

```diff
@@ -320,56 +328,59 @@ private void generateProtocolVersionCheck(MethodSpec.Builder builder) {
 
-  private boolean setEntityHasDoNotSet(ExecutableElement v) {
-    SetEntity annotation = v.getAnnotation(SetEntity.class);
-    if (annotation != null) {
-      return annotation.nullSavingStrategy() == NullSavingStrategy.DO_NOT_SET;
-    }
-    return false;
-  }
+    boolean hasDoNotSetStatementSetExplicitly =
+        preparedStatements
+            .stream()
+            .map(v -> v.methodElement)
+            .anyMatch(
+                v ->
+                    updateHasDoNotSet(v, true)
+                        || insertHasDoNotSet(v, true)
+                        || setEntityHasDoNotSet(v, true)
+                        || queryHasDoNotSet(v, true));
 
-  private boolean insertHasDoNotSet(ExecutableElement v) {
-    Insert annotation = v.getAnnotation(Insert.class);
-    if (annotation != null) {
-      return annotation.nullSavingStrategy() == NullSavingStrategy.DO_NOT_SET;
-    } else {
-      return false;
+    boolean hadSetToNullStatementSetExplicitlyForAll =
+        preparedStatements
+            .stream()
+            .map(v -> v.methodElement)
+            .filter(NullSavingStrategyValidation::isOperationWithNullSavingStrategy)
+            .allMatch(
+                v ->
+                    updateHasSetToNullExplicitly(v)
+                        || insertHasSetToNullExplicitly(v)
+                        || setEntitySetToNullExplicitly(v)
+                        || queryHasSetToNullExplicitly(v));
+
+    // if DAO level SET_TO_NULL check all underlying annotations for explicit set to DO_NOT_SET
+    // (they may override it)
+    if (hasDaoLevelSetToNullStatement(interfaceElement) && hasDoNotSetStatementSetExplicitly) {
+      addVersionCheckToBuilder(builder);
+      // if DAO level DO_NOT_SET check if all underlying override it explicitly to SET_TO_NULL
+    } else if (hasDaoLevelDoNotSetStatement(interfaceElement)
+        && !hadSetToNullStatementSetExplicitlyForAll) {
+      addVersionCheckToBuilder(builder);
+    } else if (hadDaoLevelUnset(interfaceElement) && hasDoNotSetStatement) {
+      // if DAO level annotation do not present, check method level strategy
+      // (including the default one)
+      addVersionCheckToBuilder(builder);
     }
   }
 
-  private boolean updateHasDoNotSet(ExecutableElement v) {
-    Update annotation = v.getAnnotation(Update.class);
-    if (annotation != null) {
-      return annotation.nullSavingStrategy() == NullSavingStrategy.DO_NOT_SET;
-    } else {
-      return false;
-    }
+  private void addVersionCheckToBuilder(MethodSpec.Builder builder) {
+    builder
+        .beginControlFlow(
+            "if (context.getSession().getContext().getProtocolVersion().getCode() <= $T.Version.V3)",
+            ProtocolConstants.class)
+        .addStatement(
+            "throw new IllegalArgumentException($S)",
+            "You cannot use NullSavingStrategy.DO_NOT_SET for protocol version V3.")
+        .endControlFlow()
+        .build();
```

Comment:
Yes, I think that this is a good idea to have a "normal" java code where possible (instead of generated). Changed (https://github.com/datastax/java-driver/pull/1246/commits/37aada0af2153961c75d4fb7714b7b42ffc56ee4)